app/diagnostics_test.go

```diff
@@ -273,3 +273,214 @@ func TestDiagnostics(t *testing.T) {
 		}
 	})
 }
+
+func TestRudderDiagnostics(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	th := SetupWithCustomConfig(t, func(config *model.Config) {
+		*config.PluginSettings.Enable = false
+	})
+	defer th.TearDown()
+
+	type payload struct {
+		MessageId string
+		SentAt    time.Time
+		Batch     []struct {
+			MessageId  string
+			UserId     string
+			Event      string
+			Timestamp  time.Time
+			Properties map[string]interface{}
+		}
+		Context struct {
+			Library struct {
+				Name    string
+				Version string
+			}
+		}
+	}
+
+	data := make(chan payload, 100)
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, err := ioutil.ReadAll(r.Body)
+		require.NoError(t, err)
+
+		var p payload
+		err = json.Unmarshal(body, &p)
+		require.NoError(t, err)
+
+		data <- p
+	}))
+	defer server.Close()
+
+	diagnosticID := "test-diagnostic-id-12345"
+	th.App.SetDiagnosticId(diagnosticID)
+	th.Server.initRudder(server.URL)
+
+	assertPayload := func(t *testing.T, actual payload, event string, properties map[string]interface{}) {
+		assert.NotEmpty(t, actual.MessageId)
+		assert.False(t, actual.SentAt.IsZero())
+		if assert.Len(t, actual.Batch, 1) {
+			assert.NotEmpty(t, actual.Batch[0].MessageId, "message id should not be empty")
+			assert.Equal(t, diagnosticID, actual.Batch[0].UserId)
+			if event != "" {
+				assert.Equal(t, event, actual.Batch[0].Event)
+			}
+			assert.False(t, actual.Batch[0].Timestamp.IsZero(), "batch timestamp should not be the zero value")
+			if properties != nil {
+				assert.Equal(t, properties, actual.Batch[0].Properties)
+			}
+		}
+		assert.Equal(t, "analytics-go", actual.Context.Library.Name)
+		assert.Equal(t, "3.0.0", actual.Context.Library.Version)
+	}
+
+	// Should send a client identify message
+	select {
+	case identifyMessage := <-data:
+		assertPayload(t, identifyMessage, "", nil)
+	case <-time.After(time.Second * 1):
+		require.Fail(t, "Did not receive ID message")
+	}
+
+	t.Run("Send", func(t *testing.T) {
+		testValue := "test-send-value-6789"
+		th.App.SendDiagnostic("Testing Diagnostic", map[string]interface{}{
+			"hey": testValue,
+		})
+		select {
+		case result := <-data:
+			assertPayload(t, result, "Testing Diagnostic", map[string]interface{}{
+				"hey": testValue,
+			})
+		case <-time.After(time.Second * 1):
+			require.Fail(t, "Did not receive diagnostic")
+		}
+	})
+
+	// Plugins remain disabled at this point
+	t.Run("SendDailyDiagnosticsPluginsDisabled", func(t *testing.T) {
+		th.App.sendDailyDiagnostics(true)
+
+		var info []string
+		// Collect the info sent.
+	Loop:
+		for {
```

Comment:
We can refactor this and the loop below to a common function - `collectInfo` and call it.

```
collectInfo := func(info []string) {
  t.Helper()
  for {
    select {
	case result := <-data:
	   assertPayload(t, result, "", nil)
	   info = append(info, result.Batch[0].Event)
	case <-time.After(time.Second * 1):
              return
	}
    }
  }
```

Reply:
:+1: